server/controllers/medical/patients/visits.js

```diff
@@ -178,13 +229,175 @@ function admission(req, res, next) {
     data.patient_uuid = db.bid(data.patient_uuid);
   }
 
-  const sql = `
+  const create = data.hospitalized ? createHospitalization : createVisit;
+  create(data)
+    .then(() => {
+      res.status(201).json({ uuid : visitUuid });
+    })
+    .catch(next)
+    .done();
+}
+
+function createVisit(data) {
+  const visitService = getVisitServiceQuery(data);
+  const visit = _.omit(data, ['bed', 'service']);
+  const { service } = data;
+  const sqlInsertVisit = `
     INSERT INTO patient_visit SET ?;
   `;
+  const transaction = db.transaction();
+  visit.last_service_id = service.id;
+  transaction.addQuery(sqlInsertVisit, [visit]);
+  transaction.addQuery(visitService.query, [visitService.parameters]);
+  return transaction.execute();
+}
+
+function createHospitalization(data) {
+  const visitService = getVisitServiceQuery(data);
+  const { bed, service } = data;
+  const visit = _.omit(data, ['bed', 'service']);
+
+  return Q.fcall(() => {
+    return lookupAutoAvailableBed(bed);
+  })
+    .then(b => {
+      const sqlInsertVisit = `
+        INSERT INTO patient_visit SET ?;
+      `;
+      const sqlInsertHospitalization = `
+        INSERT INTO patient_hospitalization SET ?
+      `;
+      const sqlUpdateBed = `
+        UPDATE bed SET is_occupied = 1 WHERE id = ?;
+      `;
+      const paramInsertHospitalization = {
+        uuid : db.bid(uuid()),
+        patient_visit_uuid : visit.uuid,
+        patient_uuid : visit.patient_uuid,
+        room_uuid : db.bid(bed.room_uuid),
+        bed_id : b.id,
+      };
+      const transaction = db.transaction();
+      // insert a new patient visit
+      visit.last_service_id = service.id;
+      transaction.addQuery(sqlInsertVisit, [visit]);
+      // insert a new patient visit service
+      transaction.addQuery(visitService.query, [visitService.parameters]);
+      // insert a new patient hospitalization
+      transaction.addQuery(sqlInsertHospitalization, [paramInsertHospitalization]);
+      // update the bed for the hopitalized patient
+      transaction.addQuery(sqlUpdateBed, [paramInsertHospitalization.bed_id]);
+      return transaction.execute();
+    });
+}
+
+function getVisitServiceQuery(data) {
+  const { service } = data;
+  const visit = _.omit(data, ['bed', 'service']);
+  const sqlInsertVisitService = `
+    INSERT INTO patient_visit_service SET ?;
+  `;
+  const paramInsertService = {
+    uuid : db.bid(uuid()),
+    service_id : service.id,
+    patient_visit_uuid : visit.uuid,
+  };
+  return {
+    query : sqlInsertVisitService,
+    parameters : paramInsertService,
+  };
+}
 
-  db.exec(sql, [data])
+function lookupAutoAvailableBed(bedOptions) {
+  const sql = `
+    SELECT b.id FROM bed b 
+    JOIN room r ON r.uuid = b.room_uuid
+    WHERE r.uuid = ? AND b.is_occupied = 0
+    ORDER BY r.label, b.label
+    LIMIT 1;
+  `;
+  return bedOptions.id ? { id : bedOptions.id } : db.one(sql, [db.bid(bedOptions.room_uuid)]);
```

Comment:
Hmm...  This function will either return a promise or an object.  It seems like this is a [code smell](https://en.wikipedia.org/wiki/Code_smell).  If you ever have this scenario in the future, just wrap the object in `Promise.resolve()`, like this:
```js
return bedOptions.id ? Promise.resolve({id : bedOptions.id}) : db.one(sql, [ ... ]);
```

However, it seem like if we have `bedOptions.id`, this function doesn't do anything.  But if that is the case, why did we call it in the first place?  We should have done something like this:
```js
if (options.id) {
  bed = { id : options.id };
}  else {
  bed = await lookupNextAvailableBed(options).
}
```